File: Java/Find Median from Data Stream.java
//2 Priorityqueue   O(log(N))   O(N)   
class MedianFinder {
    PriorityQueue<Integer> lo;
    PriorityQueue<Integer> hi;

    /** initialize your data structure here. */
    public MedianFinder() {
        lo = new PriorityQueue<>(new Comparator<Integer>() {
        @Override
        public int compare(Integer a1, Integer a2) {
            return a2 - a1;
        }
        });
        
        hi = new PriorityQueue<>();
    }
    
    public void addNum(int num) {
        lo.offer(num);
        hi.offer(lo.peek());
        lo.poll();
        if(lo.size() < hi.size()) {
            lo.offer(hi.peek());
            hi.poll();
        }
    }
    
    public double findMedian() {
        return lo.size() > hi.size() ? (double) lo.peek() : (lo.peek() + hi.peek())*0.5;
    }
}

File: Python/ds design/Find Median from Data Stream.py
from heapq import *
class MedianFinder(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.smaller = []
        self.larger = []
        self.median = 0

    def addNum(self, num):
        """
        :type num: int
        :rtype: void
        """
        if num > self.median:
            heappush(self.larger, num)
            if len(self.larger) > (len(self.smaller) + 1):
                a = heappop(self.larger)
                heappush(self.smaller, -a)
            self.median = self.larger[0]
        else:
            heappush(self.smaller, -num)
            if len(self.smaller) > (len(self.larger) + 1):
                a = heappop(self.smaller)
                heappush(self.larger, -a)
            self.median = -self.smaller[0]
                
        if len(self.larger) == len(self.smaller):
            self.median = float(self.larger[0] - self.smaller[0]) / 2
        
    def findMedian(self):
        """
        :rtype: float
        """
        return self.median   